torch/_prims/context.py

```diff
@@ -96,6 +97,14 @@ def __torch_function__(
             return orig_func(*args, **kwargs)
         mapping = torch_to_refs_map()
         func = mapping.get(orig_func, None)
+
+        # For torch.ops.aten.*, use registered decompositions from torch._decomp
+        # torch._decomp.decomposition_table provides a mapping from
+        # torch.ops.aten.* to torch._refs or torch._decomp.decompositions
+        # implementations.
+        if func is None and isinstance(orig_func, torch._ops.OpOverload):
```

Comment:
I have mixed feeling with this blind decomposition.

For cases where fallback is enabled, we are never leaking this out side, so we should be good with nvfuser.
But I think in general, aten2aten decomposition is not applied universally and have always been hand-picked. https://github.com/pytorch/torchdynamo/pull/584/files#diff-4b543c3fb1258c80c9be8c901c0f189ec5fdfb86ab4900313102b2481f53f063R183-R206

I'll leave this discussion to @SherlockNoMad & @Chillee

Reply:
We have torch calls to refs implemented in this file with `torch_to_refs_map` dict, `torch._decomp.decompositions` is used as mapping of ATen ops to refs.

Once the ref function is found (no matter as a torch->ref or aten->ref) then it's being queried whether to use it or not with `should_fallback_fn` callback function.

It's not unconditional blind decomposition as with [`torch.fx.experimental.proxy_tensor.decompose(decomposition_table)`](https://github.com/pytorch/pytorch/blob/14b660fcc0c32a7478c69a25a253b67d0ed36364/torch/fx/experimental/proxy_tensor.py#L77).

Reply:
Thanks for bearing with me~

Was aten calls just not translated to prim? I guess that's possible cause we just weren't using the context manager previously and have decomposition used in the other code path.

A follow up question, here looks like are are mapping `torch._decomp.decomposition_table` to all `orig_func`. How are we restricting the map to only allow `aten->ref` and not other `aten->aten` or whatever decomposition gives us?

Reply:
> Was aten calls just not translated to prim?

Yes. It's a bit confusing to have several ways to arrive at the decomposition composed of prims, I think at some point we will converge at having only `torch._refs` namespace and migrate `torch._decomp.decompositions` functions there.

`torch._decomp.decomposition_table` maps `aten` functions to functions implemented mostly in `torch._refs` (about 200 functions) but there are a few functions implemented in `torch._decomp.decompositions` (70 functions). Both `torch._refs` and `torch._decomp.decompositions` functions are implemented in terms of `torch.*` calls and are decomposed down to prims in this context manager due to this PR https://github.com/pytorch/pytorch/pull/80135.

We could restrict the map only to functions defined in `torch._refs`, but we'd miss the backward decompositions currently defined in `torch._decomp.decompositions`.

Reply:
> I think at some point we will converge at having only torch._refs namespace and migrate torch._decomp.decompositions functions there.

I don't think this is the full story, because it is ambiguous what exactly "having only a torch._refs namespace" means. And this confusion stems from a more basic confusion about what exactly TorchRefsMode is supposed to be used for (you're using it to trigger decompositions later in the stack; I think this is a misuse of the mode.) All of this confusion arises because there are a lot of ways to do what is effectively the same thing. Let me try to unpack it.

**What TorchRefsMode was originally intended to do.** Let's suppose that I have some user code that calls `torch.add()`, and I want to trace it into an FX graph of torch.ops calls, with PrimTorch decompositions applied. To be clear, this is NOT the situation that this PR is addressing. There are two ways I can do it:

* I can use TorchRefsMode, interposing directly on the Python torch.add call, directly shunting to torch._refs.add. I'll bounce around in the refs code, maybe recursively hitting TorchRefsMode for torch.* calls inside the refs, until I finally hit torch.ops.prim calls which will trace into my graph, as they do not decompose further. 
* I can NOT use TorchRefsMode, and instead rely on the decomposition table built into proxy tensor. In this case, I'll go into the traditional eager C++ binding code for torch.add, eventually coming back to `__torch_dispatch__` on proxy tensor at torch.ops.aten.add.default, and then I'll discover there's a torch.ops.aten.add mapping (from PrimTorch, because whenever possible we try to register PrimTorch decompositions to the decompositions table), and *then* I'll hit the refs code, bouncing around recursively in the same way.

Which of these is better? Well, it depends. If you use TorchRefsMode, it is tantamount to *swapping out the entire PyTorch frontend with torch._refs.* This is faithful to the long term intention of PyTorch refs: they're supposed to be a wholesale replacement of the entire PyTorch frontend for a universe where PT2 can unconditionally be turned on and there's no downside to having as much of our frontend written in Python as possible. Even if an operator is CompositeImplicitAutograd, torch._refs will get the call and be able to interpose.

This upside is also a downside: torch._refs is forced to handle ALL of PyTorch's argument parsing vagaries (and as you've seen, it often gets them wrong, because there are a lot of little special cases PyTorch has accumulated over time), and you get punished for it if you get it wrong. The calls you get in torch dispatch are a bit more regular and easier to deal with (for one, you don't have to deal with overload resolution!)

**What this PR is trying to do.** The above situation was what I had in mind when I wrote TorchRefsMode, but it's not the situation that you are doing in this PR. In this PR, you have an FX graph produced by AOT Autograd (so it's already torch.ops.aten calls), and you are trying to decompose them into a decomposed graph of torch.ops.prims calls (when nvfuser can support it), but leave it alone otherwise. Once again, there's a lot of different ways you can do it.

* What this PR does: retrace the FX graph, but using TorchRefsMode, and add extra support for handling torch.ops.aten calls (since otherwise TorchRefsMode doesn't do anything).
* Symmetrically to the two options in the previous case, you could retrace the FX graph, but use a TorchDispatchMode  to do the decompositions. This would effectively do the same thing as TorchFunctionMode, since we go from torch function immediately to torch dispatch; there's no autograd or any subsystems involved at this point. On the plus side, we don't have to jam this extra torch.ops handling into TorchRefsMode (which was originally specced as just a way to swap the Python frontend between torch and torch._refs). Note that you are welcome to use this *in conjunction* with TorchRefsMode, which can make a difference if you really, really want to shunt torch.* calls in the decompositions immediately to torch._refs, but this shouldn't matter much since you're going to be tracing into torch.ops calls anyway in the end.
* @Chillee's preferred method: don't bother retracing the entire graph, because you *already* have an FX graph, so you can just iterate through the nodes one by one and figure out if you want to decompose them or not, without having to deal with an ambient proxy tensor mode that requires you to deal with reentrant mode handling. In particular, this would have eliminated the need for https://github.com/pytorch/pytorch/pull/81764 as you wouldn't have had an ambient tracing context that you would have to disable. On the downside, you're responsible for splicing in the generated graphs from the graphs (you have to `make_fx` them) into the outer graph that you're building by hand; so this is a little bit more code.

Given that we already spent more than a week sweating over https://github.com/pytorch/pytorch/pull/81764, it probably is fine to stick with the running code we have right now (unless, maybe, you really like @Chillee's design, and the trace in a trace is still giving you headaches). But if we stick with the current code, that still leaves with a choice between the first two options.

**What to do?** I think rough consensus and running code should rule here, and the approach here isn't wrong, so given that it is all working, I'm not going to ask for more changes on top. But I want us to be cognizant of the other choices in the design space, and being open to swapping to them if it does save us a lot of grief. For example, if you're tired of spending time resolving inconsistencies in torch._refs Python argument parsing with PyTorch proper, that may be a good reason to switch to a torch dispatch decomposer, and leave TorchRefsMode as an aide for testing that PrimTorch is doing a good job in reimplementing the PyTorch frontend.

Reply:
One more thing: we can't get rid of direct to aten registrations so long as we are using C++ autograd, because there are plenty of ATen functions that show up from autograd that are not bound from the public Python API. There is no "logical" place to put them in torch._refs, because they don't exist anywhere in the public Python API. So you'll probably still have something like `torch._refs.ops.aten` to stash them even after you unify the two modules.

Reply:
> @Chillee's preferred method: don't bother retracing the entire graph, because you already have an FX graph, so you can just iterate through the nodes one by one

To clarify, this is my preferred method of writing most transformations on FX graphs. 

This approach of using `TorchRefsMode` to interject has been somewhat hard for me to reason through, since it seems to me like there are many different locations that it can be interjecting when you're using it intermixed with `__torch_dispatch__`

Some examples:

1.  Call `torch.add` => `__torch_function__` => map to refs using existing TorchRefsMode.
2. Call `torch.ops.aten.add` => `__torch_function__` => use the additional mapping in this PR to map to refs.
3. Call `aten::add` from C++ => `__torch_dispatch__` (torch.ops.aten.add) => in ProxyTensor, call `torch.ops.aten.add` from Python within `__torch_dispatch__` (is it enabled here?) => `__torch_function__` => use the additional mapping in this PR that works for `torch.ops.aten.add` to map to refs.
4. Call `torch.foo` => `__torch_function__` => TorchRefsMode (no existing ref for `torch.foo`, but `CompositeImplicitAutograd` impl exists) => `aten::foo` (CompositeImplicitAutograd) => `aten::bar` => go to step 3.

I'm sure there are other cases I haven't thought through either. 

In comparison, with `__torch_dispatch__` there's only one interjection point. Going through the above examples (here "regular PyTorch" means that we don't hit any `__torch_function__`/`__torch_dispatch__` interception.

1. Call `torch.add` => regular PyTorch => `__torch_dispatch__` (torch.ops.aten.add).
2. Call `aten.add` => regular PyTorch => `__torch_dispatch__` (torch.ops.aten.add).
3. Call `aten::add` from C++ => regular PyTorch => `__torch_dispatch__` (torch.ops.aten.add).
4. Call `torch.foo` => regular PyTorch => `__torch_dispatch__` (torch.ops.aten.bar).

Essentially, interjecting at the `__torch_dispatch__` level only unifies the interface of ops we need to care about to non-Composite ATen operators only, which is easier for me to reason about.

Reply:
Examples 3 and 4 are not working with the current setup of TorchRefsMode.

I think you know that ProxyTensor and ProxyTorchDispatchMode use `__torch_function__` within `__torch_dispatch__`, also intermixed: here
https://github.com/pytorch/pytorch/blob/780a303a2a2b865e821542f0251a0166f36d8d19/torch/fx/experimental/proxy_tensor.py#L164
Proxy's `__torch_function__` override mechanism is used and everything is fine.

> Chillee's preferred method: don't bother retracing the entire graph, because you already have an FX graph, so you can just iterate through the nodes one by one

I don't have anything against iterating through the nodes and applying transformations, but when I want to expand (decompose) a particular node I need Tensor arguments. We can use `fx.Interpreter` to modify how the nodes are run, then for code reuse, I think, it's better combined with a context manager, as DecompositionInterpreter does
https://github.com/pytorch/pytorch/blob/780a303a2a2b865e821542f0251a0166f36d8d19/torch/fx/experimental/proxy_tensor.py#L490-L496
and since all the code generated from FX graph is in Python we can consolidate the logic we need in a context manager using the Python function override mechanism.
I don't think retracing the Python code generated by the FX graph, transforming the FX graph, or reinterpreting using `fx.Interpreter` are all competing approaches, one can be implemented in terms of another.

Maybe it's a mistake on my part to rely on the existing code and slightly alter it to do things I need.
Ed's `TorchRefsMode` and Chillee's `make_fx` work well together to produce an FX graph with prims if the function being traced is the original one consisting of Python torch calls. Since `make_fx` produces an FX graph with `torch.ops.aten` calls and using `torch.ops.aten` is a valid Python code I thought it would be fine to amend the applicability of `TorchRefsMode` to cover these calls as well. So that the result of `make_fx` under `TorchRefsMode` is the same because it's just two different ways of calling the same operation.
```py
f1 = lambda x: torch.ops.aten.sum.default(x)
f2 = lambda x: torch.sum(x)
with TorchRefsMode():
    make_fx(f1)(a).graph.print_tabular()
# opcode         name                          target                              args                                    kwargs
# -------------  ----------------------------  ----------------------------------  --------------------------------------  --------
# placeholder    x_1                           x_1                                 ()                                      {}
# call_function  convert_element_type_default  prims.convert_element_type.default  (x_1, torch.float32)                    {}
# call_function  sum_default                   prims.sum.default                   (convert_element_type_default, [0, 1])  {}
# output         output                        output                              (sum_default,)                          {}
# and the same output is for f2 with modified TorchRefsMode in this PR
```
I need this because it can be often the case that before we get into the "to prims" decomposition business the original PyTorch code could already be traced and modified and the FX graph we get contains only `torch.ops.aten` calls.

Thank you, Ed, for writing down different ways. I think the third option (FX transformation option) would share the logic and be best implemented in conjunction with other options. So let's leave it aside for now. Retracing the generated Python code is easier for me to reason about (and I have nothing against the transformation or interpreter being implemented in the future). If it's not okay to expand the scope of TorchRefsMode I can add an argument to opt-in for this new behavior or create a new "AtenRefsMode".
I haven't felt the need yet to have an "AtenRefsMode" based on TorchDispatchMode.

Reply:
> There is no "logical" place to put them in torch._refs, because they don't exist anywhere in the public Python API. So you'll probably still have something like `torch._refs.ops.aten` to stash them even after you unify the two modules.

Sure, I think "aten" submodule within `torch._refs` is a logical place. Better than two overlapping in purpose higher modules `_decomp.decompositions` and `_refs`.

> One more thing: we can't get rid of direct to aten registrations so long as we are using C++ autograd, because there are plenty of ATen functions that show up from autograd that are not bound from the public Python API. 

Right, as long as we are using C++ autograd rules we need to reimplement lots of "_backward" ATen functions . `torch._decomp.decompositions` has 33 backward registrations. My thinking is that if we had derivative rules implemented for prims then we wouldn't need backward decompositions/references.

Reply:
> Examples 3 and 4 are not working with the current setup of TorchRefsMode.

Yes, this is a problem, but it should be solved directly. After attempting and failing to turn on CompositeImplicitAutograd by default for Python, I decided to solve it a different way in https://github.com/pytorch/pytorch/pull/83012

> Proxy's __torch_function__ override mechanism is used and everything is fine.

Sure. Though, actually, I'd quite like it if we didn't do this (and manually constructed the nodes). Because sometimes we disable TorchFunction and then this doesn't work XD https://github.com/pytorch/pytorch/pull/82647

> I don't have anything against iterating through the nodes and applying transformations, but when I want to expand (decompose) a particular node I need Tensor arguments.

Yes. But in fact, you do have Tensor arguments: after running FakeTensorProp, fake tensors are stored on `meta['fake_result']` for all nodes. You can use these to initialize the `make_fx` call. Is this a pain to do today? Yes,  it requires a bit of low level mucking around in FX. So I don't blame you for not instantly knowing you could do this ;)

I do think retracing the entire forward is more obvious. It just has a *big* downside, which is that doing raw Graph manipulation is very complicated. I went through a similar process trying a lot of different ways to write the cuda graphs graph transformation; I was able to get the raw graph to work, but it was similarly impenetrable.

> I don't think retracing the Python code generated by the FX graph, transforming the FX graph, or reinterpreting using fx.Interpreter are all competing approaches

Well, they are competing in the sense that, for any given pass we want to implement, we're going to implement it one way, not all three (equivalent) ways lol.

> My thinking is that if we had derivative rules implemented for prims then we wouldn't need backward decompositions/references.

This is a long way away! haha

Reply:
> My thinking is that if we had derivative rules implemented for prims then we wouldn't need backward decompositions/references.

I think this isn't so difficult. There are only 63 backwards ops total, and we already have decompositions for 34 of them. 

Basically, imo, there are 2 main ways we can get everything to prims.

1. Decompose into prims upfront, and then perform all transformations (i.e. autograd) on top of prims. 
2. Trace through, and then decompose the resulting ATen operators we see.

There are different decompositions we can avoid writing for each approach.

Pros of 1: 
- We don't need to write decompositions for "private" operators (i.e. `_log_softmax_backward_data`).
- We don't need to write decompositions for "backwards" operators.

Pros of 2:
- We don't need to write decompositions for "CompositeImplicitAutograd" operators (i.e. `at::einsum`) 
- We don't need to write autograd formulas.

Which one is less work? Well, I think ... 2. Concretely, there are 485 non-composite Aten operators tested by OpInfo tests today (https://docs.google.com/spreadsheets/d/15UhFgTLojCcu4A-2NONIhu6etxyd1WZn0K7U-Yk8KAs/edit#gid=103957467). OTOH, there are 514 public PyTorch operators tested by OpInfo tests today (https://docs.google.com/spreadsheets/d/15UhFgTLojCcu4A-2NONIhu6etxyd1WZn0K7U-Yk8KAs/edit#gid=2147055876). 

Note that the 485 non-composite ATen operators also contains 63 backwards operators and 27 private operators. 

But doing 1 also requires us to add support for batching rules/reverse-mode AD/forward-mode AD formulas for all prims (at a minimum). Overall, in my opinion, I think 2 is less work, although I think it's reasonable to believe 1 is less work.

That being said, in some sense, this discussion is purely academic. We've settled on approach 2 for the short-term (cc: @ezyang, @ngimel if folks disagree). There are other advantages/disadvantages of each approach (and we might move in a different direction in the long term), but approach 2 imposes the least requirements on the rest of our system.

Reply:
Approach 1 amounts to writing references/decompositions for `torch` functions, while approach 2 is about writing references/decompositions for `aten` "CompositeExplicitAutograd" functions. There's some overlap, yes, but not much, and differences in the arguments could slightly complicate things further. In the linked spreadsheet it's stated that having 100% coverage of `torch` refs would give 40% coverage for `aten` refs.

> We've settled on approach 2 for the short-term
approach 2 imposes the least requirements on the rest of our system.

That's very helpful that you stated this. Does it mean that `torch` decompositions have no value for you and others now and are not a priority in the short term?

> Basically, imo, there are 2 main ways we can get everything to prims.

I see also a third way because we don't need to get everything to prims in the short-term (and probably neither in the long-term): Decompose into prims upfront, *only where needed and useful*, and then support different transformations for the combination of prims and aten ops, at least `.backward()`.

Reply:
> There's some overlap, yes, but not much, and differences in the arguments could slightly complicate things further.

There's actually quite a bit of overlap I think. The spreadsheet isn't super clear about this, but the "true progress" is already accounting for all of the refs that are already registered as decompositions (so even the most basic accounting gives 66% coverage). In addition, the estimate of "overlap" is quite pessimistic - many of the aten operators that don't directly map to an identical name still map to some functionality in the public API.

For example, `slice.Tensor`, `im2col`, `_fft_r2c` maps directly to a prim, etc. The only functionality that is "truly" potentially redundant is the backwards ops, and there's only 63 of them (of which we've already decomposed 34).

I do agree that differences in arguments are somewhat annoying to deal with, but seem manageable to me.

>  Does it mean that torch decompositions have no value for you and others now and are not a priority in the short term?

So, first of all, if a torch op has the same functionality as an aten op, it's obviously a priority. In addition, in general, I think it's nice to have functionality implemented in Python, including for `CompositeImplicitAutograd` ops.

However, from a practical perspective, yes, I think implementations of ops that are only in the torch API, or are `CompositeImplicitAutograd` are a lower priority. So, for example, I don't think reference implementations of `einsum`/`instance_norm` are a priority (once again, cc: @ezyang and @ngimel if they want to add any additional context/disagree).

> Decompose into prims upfront, only where needed and useful, and then support different transformations 

We discussed an approach like this to some extent a while back, but decided against it in the short term for a couple reasons (well, at least, these are my main reasons offhand)

1. This makes it harder to test individual components of our system for faithfulness against PyTorch eager. For example, today it's possible to test the autograd/functionalization/other functionality capture without needing to also use decompositions at all. However, if we rely upon differentiating all of our relevant operators up front this is no longer true - our transformations are fundamentally tied up with our decompositions.
2. It's a significant amount of work to support all of our transformations on top of the prims, and we're not even totally sure about all of the ramifications. One thing is that it seems pretty likely to me that we need a way of preserving higher-level structure. As an example, how is amp gonna deal with layer norm after we've decomposed it into its constituent components? 
3. I don't see a significant (short-term) advantage to doing so. In the long run, I can see the appeal of having a small primitive set of operators we write our transforms over, but this doesn't immediately help us in the short term.
4. Decomposing into prims up-front makes it harder for backends to control what ops they see. For example, let's say we're using the "speculative decomposition" approach in this PR. How do we decide whether `torch.foo` should be decomposed into prims that we then transform? Not only do we need to check whether `torch.foo` itself can be decomposed into supported prims, we also need to check whether the result of `torch.foo_backward`, `torch.foo_vmap`, or whatever other combination of transforms can *also* be decomposed into prims

Reply:
> How do we decide whether torch.foo should be decomposed into prims that we then transform? Not only do we need to check whether torch.foo itself can be decomposed into supported prims, we also need to check whether the result of torch.foo_backward, torch.foo_vmap, or whatever other combination of transforms can also be decomposed into prims

Imagine that all supported prims support all interesting transforms. Once `torch.foo` was decided to be decomposed into prims there's no need to check whether `foo_backward`, `foo_vmap`, etc. is decomposable into supported prims, we would use the primal function with prims in it and use them in the transformations. Maybe there is no interest in working towards that future.

Reply:
> Once torch.foo was decided to be decomposed into prims there's no need to check whether foo_backward, foo_vmap, etc. is decomposable into supported prims, we would use the primal function with prims in it and use them in the transformations. Maybe there is no interest in working towards that future.

The problem is that if you *do* have the transformation for foo, you're generally not going to be willing to drop it, because in conventional eager mode you don't actually want to decompose since that will destroy fusion. In a hypothetical universe where a compiler is always on and you can always recover perf from this situation, this is fine to do unconditionally, but hopefully it's clear that *that* is going to a harder place to get to, and further in the future.

Reply:
> Once torch.foo was decided to be decomposed into prims there's no need to check whether foo_backward, foo_vmap, etc. is decomposable into supported prims

What if your backend is able to recover perf from decomposing `torch.foo`, but not from decomposing `torch.foo_backward`? 

As an easy example, you could probably decompose `torch.dot` into `prims.mul` and `prims.sum` without any issue. But, if we vmap `torch.dot`, that turns into a matmul, and you don't want to decompose that.